samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -231,21 +237,33 @@ public String getNextOpId(OpCode opCode) {
    * @return  the intermediate {@link MessageStreamImpl}
    */
   @VisibleForTesting
-  <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
+  public <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
     Preconditions.checkState(!inputOperators.containsKey(streamId) && !outputStreams.containsKey(streamId),
         "getIntermediateStream must not be called multiple times with the same streamId: " + streamId);
 
     if (serde == null) {
-      LOGGER.info("Using default serde for intermediate stream: " + streamId);
-      serde = (Serde<M>) defaultSerde;
+      LOGGER.info("No serde provided for intermediate stream: " + streamId +
+          ". Key and message serdes configured for the job.default.system will be used.");
     }
 
     if (isBroadcast) broadcastStreams.add(streamId);
-    boolean isKeyed = serde instanceof KVSerde;
-    KV<Serde, Serde> kvSerdes = getKVSerdes(streamId, serde);
+
+    boolean isKeyed;
+    KV<Serde, Serde> kvSerdes;
+    if (serde == null) { // if no explicit serde available
+      isKeyed = true; // assume keyed stream
+      kvSerdes = new KV<>(null, null); // and that key and msg serdes are provided for job.default.system in configs
+    } else {
+      isKeyed = serde instanceof KVSerde;
+      kvSerdes = getKVSerdes(streamId, serde);
+    }
+
+    InputTransformer transformer = (InputTransformer) defaultSystemDescriptorOptional
```

Comment:
nit: the type cast is redundant.

Reply:
It's actually required, since otherwise it infers the type to be Object (common super type of InputTransformer and null)